### PR TITLE
Fix repl discarding --build-depends

### LINKED ISSUE #6859: --build-depends with repl does not consume its list of inputs if it sees allow-newer
**UPD 2023-02-03** Originally, the bug report suggested that the bug is always there, with any project file. It was later discovered by @wismill that the bug only manifests if cabal sees `allow-newer` -- either in the project file or in the command-line arguments.

**Describe the bug**
The `--build-depends` flag does not seem to be respected within a project dir, but oddly works outside of it. I'd expect the following to bring the list of dependencies into scope for the repl for a project, but it does not: 

```haskell
λ П(a: A) base64-bytestring → λ git emily/check-decode-padding → cabal repl --build-depends "vector, sequence, containers, unordered-containers" 
Resolving dependencies...
Build profile: -w ghc-8.8.3 -O1
...
Ok, five modules loaded.
П> import qualified Data.Map as M

<no location info>: error:
    Could not load module ‘Data.Map’
    It is a member of the hidden package ‘containers-0.6.2.1’.
    Perhaps you need to add ‘containers’ to the build-depends in your .cabal file.
```

However, if i step outside of the project directory, it works: 

```haskell
λ П(a: A) base-encodings → cabal v2-repl --build-depends "vector, sequence, containers, unordered-containers"
Resolving dependencies...
Build profile: -w ghc-8.8.3 -O1
In order, the following will be built (use -v for more details):
 - fake-package-0 (lib) (first run)
Configuring library for fake-package-0..
Preprocessing library for fake-package-0..
Warning: No exposed modules
GHCi, version 8.8.3: https://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /Users/emilypi/.ghc/ghci.conf
Loaded GHCi configuration from /var/folders/64/kt1f18kn7kzfhjgw25cf1r140000gn/T/cabal-repl.-24971/setcwd.ghci
П> import qualified Data.Map as M
П> 
```

**To Reproduce**

Run the above commands. 

**Expected behavior**

I would expect the behavior of `--build-depends` to be concatenative (modulo conflicts), and work the same inside of a project directory as well as outside. 

**System information**

- OS: MacOS
- GHC: 8.8.3
- Cabal: 3.2

### LINKED ISSUE #7081: cabal-install >= 3.4 regression: `cabal repl -b` fails to install packages in combination with `--allow-newer`
To reproduce this, pick a package that you haven't installed yet, and try the following with `cabal-install-3.4`:

```
$ /opt/cabal/3.4/bin/cabal --version
cabal-install version 3.4.0.0
compiled using version 3.4.0.0 of the Cabal library

$ /opt/cabal/3.4/bin/cabal repl -b distributive -w /opt/ghc/8.10.2/bin/ghc --allow-newer=base
Resolving dependencies...
Build profile: -w ghc-8.10.2 -O1
In order, the following will be built (use -v for more details):
 - fake-package-0 (lib) (first run)
Configuring library for fake-package-0..
Preprocessing library for fake-package-0..
```

Notice that `distributive` isn't built at all! Note that the use of the `distributive` and `base` packages aren't important here. You can substitute them with any other packages and you should still observe a similar effect.

The bug does not occur if:

* `--allow-newer` is omitted:

  ```
  $ /opt/cabal/3.4/bin/cabal repl -b distributive -w /opt/ghc/8.10.2/bin/ghc
  Resolving dependencies...
  Build profile: -w ghc-8.10.2 -O1
  In order, the following will be built (use -v for more details):
   - base-orphans-0.8.2 (lib) (requires build)
   - cabal-doctest-1.0.8 (lib) (requires build)
   - tagged-0.8.6 (lib) (requires build)
   - distributive-0.6.2 (lib:distributive) (requires build)
   - fake-package-0 (lib) (first run)
  ```

* You use `install` instead of `repl -b`:

  ```
  $ /opt/cabal/3.4/bin/cabal install distributive -w /opt/ghc/8.10.2/bin/ghc --allow-newer=base
  Resolving dependencies...
  Build profile: -w ghc-8.10.2 -O1
  In order, the following will be built (use -v for more details):
   - base-orphans-0.8.2 (lib) (requires build)
   - cabal-doctest-1.0.8 (lib) (requires build)
   - tagged-0.8.6 (lib) (requires build)
   - distributive-0.6.2 (lib:distributive) (requires build)
  ```

* You use `cabal-install-3.2` instead of `cabal-install-3.4`:

  ```
  $ /opt/cabal/3.2/bin/cabal --version
  cabal-install version 3.2.0.0
  compiled using version 3.2.0.0 of the Cabal library
  
  $ /opt/cabal/3.2/bin/cabal repl -b distributive -w /opt/ghc/8.10.2/bin/ghc --allow-newer=base
  Resolving dependencies...
  Build profile: -w ghc-8.10.2 -O1
  In order, the following will be built (use -v for more details):
   - base-orphans-0.8.2 (lib) (requires build)
   - cabal-doctest-1.0.8 (lib) (requires build)
   - tagged-0.8.6 (lib) (requires build)
   - distributive-0.6.2 (lib:distributive) (requires build)
   - fake-package-0 (lib) (first run)
  ```

### PR DESCRIPTION
**Pull request by wismill**
Fix REPL discarding `--build-depends` when using `allow-newer` or `allow-older`.

Fixes #6859
Fixes #7081
Related issues: #8504, #5900.

---

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by wismill** [comment]
Note: this is currently hacky. It fixes my use case (namely run doctest), but I am not sure it fixes it in general.

The following points need review:
- The combination of dependencies lists in `mkCondNode`: currently done naively.
- The hack on `condTreeConstraints` field: we have a function `g :: BuildInfo → f BuildInfo` but we need to update `[Dependency]`.

File: Cabal-syntax/src/Distribution/Types/GenericPackageDescription.hs [hunk 1]
```diff
@@ -86,19 +85,24 @@ instance L.HasBuildInfos GenericPackageDescription where
...
     => LensLike' f (CondTree v [Dependency] comp) L.BuildInfo
 traverseCondTreeBuildInfo g = node where
-    mkCondNode :: comp -> [CondBranch v [Dependency] comp] -> CondTree v [Dependency] comp
-    mkCondNode comp branches = CondNode comp (view L.targetBuildDepends comp) branches
+    mkCondNode
+      :: comp
+      -> [Dependency]
+      -> [CondBranch v [Dependency] comp]
+      -> CondTree v [Dependency] comp
+    mkCondNode comp deps = CondNode comp (view L.targetBuildDepends comp <> deps)
```
Comment (wismill) [inline Cabal-syntax/src/Distribution/Types/GenericPackageDescription.hs:99]: This naive concatenation makes the test suite fail, as I suspected. I do think the issue is with this field, but it may be due to its misuse before the call to `traverseCondTreeBuildInfo`.

**Comment by wismill** [comment]
New version, which hopefully is correct and much simplier.

**Comment by ulysses4ever** [comment]
I hope to attend to this PR ASAP. I really hope we can do squeeze it into 3.10 @Mikolaj

File: cabal-testsuite/PackageTests/ReplBuildDepends/cabal.test.hs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import Test.Cabal.Prelude
+
+-- Test “repl --build-depends”
```
Comment (ulysses4ever) [inline cabal-testsuite/PackageTests/ReplBuildDepends/cabal.test.hs:3]: Looks like the script repeats the same 7 lines three times, and the only difference is the extra argument (none, or `allow-newer`, or `allow-older`). Could you try to refactor the identical part into a separate function in this file?

File: Cabal-syntax/src/Distribution/Types/GenericPackageDescription.hs [hunk 3]
```diff
@@ -86,15 +85,14 @@ instance L.HasBuildInfos GenericPackageDescription where
...
-      where
 
 -- We use this traversal to keep [Dependency] field in CondTree up to date.
 traverseCondTreeBuildInfo
     :: forall f comp v. (Applicative f, L.HasBuildInfo comp)
     => LensLike' f (CondTree v [Dependency] comp) L.BuildInfo
 traverseCondTreeBuildInfo g = node where
     mkCondNode :: comp -> [CondBranch v [Dependency] comp] -> CondTree v [Dependency] comp
-    mkCondNode comp branches = CondNode comp (view L.targetBuildDepends comp) branches
+    mkCondNode comp = CondNode comp (view L.targetBuildDepends comp)
```
Comment (ulysses4ever) [inline Cabal-syntax/src/Distribution/Types/GenericPackageDescription.hs:95]: Are all three changes in this files mere cleanups? Please create a separate commit for it.

File: cabal-install/src/Distribution/Client/CmdRepl.hs [hunk 4]
```diff
@@ -338,8 +338,8 @@ addDepsToProjectTarget deps pkgId ctx =
       | packageId pkg /= pkgId = SpecificSourcePackage pkg
       | SourcePackage{..} <- pkg =
         SpecificSourcePackage $ pkg { srcpkgDescription =
-          srcpkgDescription & (\f -> L.allCondTrees $ traverseCondTreeC f)
-                            %~ (deps ++)
+          srcpkgDescription & L.traverseBuildInfos
+                            %~ over L.targetBuildDepends (deps ++)
```
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/CmdRepl.hs:342]: Using `over` and its operator synonym `%~` one after another is a bit confusing. Is this the same as:
```
srcpkgDescription
  &  L.traverseBuildInfos
  %  L.targetBuildDepends 
  %~ (deps ++)
```
(compose the lenses first, and apply `over` / `%~` last) ? If yes, I'd prefer that. I'm not sure my version typechecks and/or does the same thing, but you get the idea.

**Review by ulysses4ever — CHANGES_REQUESTED** [review]
This looks great. Two requests (apart from minor inline comments):

1. A changelog file
2. A better commit sequence. Please, force-push your branch so that it doesn't have the previous attempt, and has the cleanup work in a separate commit.

**Comment by Mikolaj** [comment]
> I hope to attend to this PR ASAP. I really hope we can squeeze it into 3.10 @Mikolaj

Yes, we are only in semi-freeze, so if there are good reasons, we should merge it into 3.10. E.g., improving CI is a good reason, because 3.10 is going to stay there longer, since we need some other changes before release.

File: cabal-install/src/Distribution/Client/CmdRepl.hs (hunk 4 above)
Comment (wismill) [reply cabal-install/src/Distribution/Client/CmdRepl.hs]: Done

File: Cabal-syntax/src/Distribution/Types/GenericPackageDescription.hs (hunk 3 above)
Comment (wismill) [reply Cabal-syntax/src/Distribution/Types/GenericPackageDescription.hs:95]: Done

File: cabal-testsuite/PackageTests/ReplBuildDepends/cabal.test.hs (hunk 2 above)
Comment (wismill) [reply cabal-testsuite/PackageTests/ReplBuildDepends/cabal.test.hs:3]: Done

**Comment by wismill** [comment]
Fixed based on feedback, rebased and squashed.

**Comment by ulysses4ever** [comment]
@wismill amazing, thanks! One more thing: could you, please, explain in the commit message (1) what this change does, in plain English, as much as possible, and (2) why you think this fixes the issue?

File: cabal-install/src/Distribution/Client/CmdRepl.hs [hunk 5]
```diff
@@ -338,7 +338,7 @@ addDepsToProjectTarget deps pkgId ctx =
       | packageId pkg /= pkgId = SpecificSourcePackage pkg
       | SourcePackage{..} <- pkg =
         SpecificSourcePackage $ pkg { srcpkgDescription =
-          srcpkgDescription & (\f -> L.allCondTrees $ traverseCondTreeC f)
+          srcpkgDescription &  (L.traverseBuildInfos . L.targetBuildDepends)
```
Comment (Mikolaj) [inline cabal-install/src/Distribution/Client/CmdRepl.hs:341]: And/or a couple of lines of code comments here? Whichever you prefer. The explanation in the commit would be historical "previously there was L.allCondTrees $ traverseCondTreeC, but it was bad and now it's good", while the comment here would be ahistorical: "targetBuildDepends makes sense here because reasons, and traverseBuildInfos then completes the wondrous transformation, yielding this and that".

Comment (wismill) [reply cabal-install/src/Distribution/Client/CmdRepl.hs:346]: Done. Do you think it explains adequately?

**Comment by wismill** [comment]
@ulysses4ever I added proper comment & commit message. Unfortunately there is an issue withe the recent release of `happy`.

**Comment by Mikolaj** [comment]
Yes, our colleague has already reported the happy problem and we count on them fixing it ASAP so that we don't have to exclude the happy version ourselves. Happy times.

File: cabal-install/src/Distribution/Client/CmdRepl.hs (hunk 5 above)
Comment (Mikolaj) [reply cabal-install/src/Distribution/Client/CmdRepl.hs:346]: 🥇

**Comment by wismill** [comment]
Note that I am new to Cabal and although I think the patch is consistent, please double check it!

There might be more places where the pattern `L.allCondTrees . traverseCondTreeC` might be an issue, e.g. `transformAllBuildDepends`.

**Review by ulysses4ever — APPROVED** [review]
Typo in the commit message: "manipuling". Other than that — terrific, thanks!

**Comment by Mikolaj** [comment]
> There might be more places where the pattern `L.allCondTrees . traverseCondTreeC` might be an issue, e.g. `transformAllBuildDepends`.

Please kindly open a ticket about that. We don't have to do everything in one go.

**Comment by Mikolaj** [comment]
> Note that I am new to Cabal and although I think the patch is consistent, please double check it!

Given we want to backport this into 3.10, let's ask our resident gurus @grayjay, @gbaz and @fgaz for a quick look.

File: cabal-install/src/Distribution/Client/CmdRepl.hs [hunk 6]
```diff
@@ -75,7 +75,7 @@ import Distribution.Types.BuildInfo
 import Distribution.Types.ComponentName
          ( componentNameString )
 import Distribution.Types.CondTree
-         ( CondTree(..), traverseCondTreeC )
```
Comment (fgaz) [inline cabal-install/src/Distribution/Client/CmdRepl.hs:78]: This is a breaking change and can't be backported as is

**Review by gbaz — APPROVED** [review]

File: cabal-testsuite/PackageTests/ReplBuildDepends/cabal.allow-newer.out [hunk 7]
```diff
@@ -0,0 +1,2 @@
+# cabal clean
+# cabal repl
```
Comment (grayjay) [inline cabal-testsuite/PackageTests/ReplBuildDepends/cabal.allow-newer.out:2]: Are `cabal.allow-newer.out` and `cabal.allow-older.out` still needed?

Comment (wismill) [reply cabal-testsuite/PackageTests/ReplBuildDepends/cabal.allow-newer.out]: It seems not. Good catch!

Comment (wismill) [reply cabal-testsuite/PackageTests/ReplBuildDepends/cabal.allow-newer.out:2]: Well, yes they were supposed to… I just realise I gave the same name to each individual test. Fixed now. But yes the `*.out` files are identical.

**Comment by Mikolaj** [comment]
@wismill: feel free to set a lable (squash+merge_me I guess), which will still give @fgaz 2 days to ponder and grace us with a comment, if @fgaz so chooses.

BTW, let's not forget to open a ticket for the other occurrences of the same pattern in the codebase. Even if we later close the ticket with "nothing to do".

**Comment by Mikolaj** [comment]
@mergify rebase

**Comment by ulysses4ever** [comment]
Regarding backport to 3.10: I'm of two minds here. On one hand, I truly believe this can make 3.10 release greater and brighter. On the other, I think it may break some stuff, perhaps — at least, I didn't see anyone on this thread playing that it can't. So, perhaps, it's Mikolaj's call.

**Comment by Mikolaj** [comment]
@ulysses4ever: I asked people at the last devs meeting to give this an extra scrutiny specifically due to the late semi-freeze backporting and we did get a lot of reviews in. However, I find your sudden lack of faith disturbing. :)

**Comment by Mikolaj** [comment]
@wismill: do we have the ticket about "the other occurrences of the same wrong pattern in our codebase"? If so, it would be useful to link it here for cross-referencing.

**Comment by Mikolaj** [comment]
@mergify backport 3.10
